pfrl/agents/hrl/hiro_agent.py

```diff
@@ -0,0 +1,252 @@
+        and train both the low and high level controllers.
+        """
+
+        if global_step >= start_training_steps:
+            # start training once the global step surpasses
+            # the start training steps
+            self.low_con.observe(obs, subgoal, self.sr, done)
+
+            if global_step % self.train_freq == 0 and len(self.action_arr) == self.train_freq:
+                # train high level controller every self.train_freq steps
+                self.high_con.agent.update_high_level_last_results(self.last_high_level_obs, self.last_high_level_goal, self.last_high_level_action)
+                self.high_con.observe(self.low_con, self.state_arr, self.action_arr, self.cumulative_reward, goal, obs, done)
+                self.action_arr = []
+                self.state_arr = []
+
+                # reset last high level obs, goal, action
+                self.last_high_level_obs = torch.FloatTensor(obs)
+                self.last_high_level_goal = torch.FloatTensor(goal)
+                self.last_high_level_action = subgoal
+                self.cumulative_reward = 0
+
+            elif global_step % self.train_freq == 0 and self.last_high_level_obs is None:
+                self.last_high_level_obs = torch.FloatTensor(obs)
+                self.last_high_level_goal = torch.FloatTensor(goal)
+                self.last_high_level_action = subgoal
+
+            self.action_arr.append(self.last_action)
+            self.state_arr.append(self.last_obs)
+            self.cumulative_reward += (self.reward_scaling * reward)
+
+    def _choose_subgoal(self, step, s, sg, n_s, goal):
+        """
+        chooses the next subgoal for the low level controller.
+        """
+        if step % self.buffer_freq == 0:
+            sg = self.high_con.policy(s, goal)
+        else:
+            sg = self.subgoal_transition(s, sg, n_s)
+
+        return sg
+
+    def subgoal_transition(self, s, sg, n_s):
+        """
+        subgoal transition function, provided as input to the low
+        level controller.
+        """
+        return s[:sg.shape[0]] + sg - n_s[:sg.shape[0]]
+
+    def low_reward(self, s, sg, n_s):
+        """
+        reward function for low level controller.
+        """
+        abs_s = s[:sg.shape[0]] + sg
+        return -np.sqrt(np.sum((abs_s - n_s[:sg.shape[0]])**2))
+
+    def end_step(self):
+        """
+        ends a step within an episode.
+        """
+        self.episode_subreward += self.sr
+        self.sg = self.n_sg
+
+    def end_episode(self, episode, logger=None):
+        """
+        ends a full episode.
+        """
+        if logger:
+            # log
+            logger.write('reward/Intrinsic Reward', self.episode_subreward, episode)
+
+            # Save Model
+            if _is_update(episode, self.model_save_freq):
+                self.save(episode=episode)
+
+        self.episode_subreward = 0
+        self.sr = 0
+
+    def save(self, episode):
```

Comment:
Is it compatible with https://github.com/HIRO-group/pfrl/blob/5aff882fce594ec46c4c0a45ae4dad71650693ee/pfrl/experiments/evaluator.py#L349 ?

Reply:
it does create a directory (that has a name with more than `episode`, so it should work, I think